File: 백준/Silver/17103. 골드바흐 파티션/골드바흐 파티션.py
import sys

def sieve(n: int):
    is_prime = bytearray(b'\x01') * (n + 1)
    if n >= 0: is_prime[0] = 0
    if n >= 1: is_prime[1] = 0
    limit = int(n ** 0.5)
    for i in range(2, limit + 1):
        if is_prime[i]:
            step = i
            start = i * i
            is_prime[start:n+1:step] = b'\x00' * (((n - start) // step) + 1)
    return is_prime

def main():
    input = sys.stdin.readline
    t = int(input().strip())
    nums = [int(input().strip()) for _ in range(t)]
    max_n = max(nums) if nums else 0

    is_prime = sieve(max_n)

    out = []
    for n in nums:
        cnt = 0
        half = n // 2
        for p in range(2, half + 1):
            if is_prime[p] and is_prime[n - p]:
                cnt += 1
        out.append(str(cnt))
    sys.stdout.write("\n".join(out))

if __name__ == "__main__":
    main()

File: 백준/Silver/17103. 골드바흐 파티션/골드바흐 파티션.cc
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int t;
    cin >> t;
    vector<int> nums(t);
    int max_n = 0;
    for (int i = 0; i < t; i++) {
        cin >> nums[i];
        max_n = max(max_n, nums[i]);
    }

    vector<char> is_prime(max_n + 1, 1);
    if (max_n >= 0) is_prime[0] = 0;
    if (max_n >= 1) is_prime[1] = 0;

    for (int i = 2; 1LL * i * i <= max_n; i++) {
        if (is_prime[i]) {
            for (int j = i * i; j <= max_n; j += i) {
                is_prime[j] = 0;
            }
        }
    }

    for (int n : nums) {
        int half = n / 2;
        int cnt = 0;
        for (int p = 2; p <= half; p++) {
            if (is_prime[p] && is_prime[n - p]) cnt++;
        }
        cout << cnt << "\n";
    }
    return 0;
}
